lib/srv/regular/sshserver_test.go

```
@@ -1148,6 +1149,228 @@ func (s *SrvSuite) TestGlobalRequestRecordingProxy(c *C) {
+	c.Assert(err, IsNil)
+
+	// Create host key and certificate for node.
+	keys, err := s.server.Auth().GenerateServerKeys(auth.GenerateServerKeysRequest{
+		HostID:               "raw-node",
+		NodeName:             "raw-node",
+		Roles:                teleport.Roles{teleport.RoleNode},
+		AdditionalPrincipals: []string{hostname},
+		DNSNames:             []string{hostname},
+	})
+	c.Assert(err, IsNil)
+
+	signer, err := sshutils.NewSigner(keys.Key, keys.Cert)
+	c.Assert(err, IsNil)
+
+	// configure a server which allows any client to connect
+	cfg := ssh.ServerConfig{
+		NoClientAuth: true,
+		PasswordCallback: func(conn ssh.ConnMetadata, password []byte) (*ssh.Permissions, error) {
+			return &ssh.Permissions{}, nil
+		},
+		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			return &ssh.Permissions{}, nil
+		},
+	}
+	cfg.AddHostKey(signer)
+
+	listener, err := net.Listen("tcp", ":0")
+	c.Assert(err, IsNil)
+
+	_, port, err := net.SplitHostPort(listener.Addr().String())
+	c.Assert(err, IsNil)
+
+	addr := net.JoinHostPort(hostname, port)
+
+	return &rawNode{
+		listener: listener,
+		cfg:      cfg,
+		addr:     addr,
+	}
+}
+
+// startX11EchoServer starts a fake node which, for each incomging SSH connection, accepts an
+// X11 forwarding request and then dials a single X11 channel which echoes all bytes written
+// to it.  Used to verify the behavior of X11 forwarding in recording proxies.
+func (s *SrvSuite) startX11EchoServer(ctx context.Context, c *C) *rawNode {
+	node := s.newRawNode(c)
+	ctx, cancel := context.WithCancel(ctx)
+	go func() {
+		defer cancel()
+		for {
+			conn, chs, _, err := node.accept()
+			if err != nil {
+				log.Warnf("X11 echo server closing: %v", err)
+				return
+			}
+			go func() {
+				defer conn.Close()
+
+				// expect client to open a session channel
+				var nch ssh.NewChannel
+				select {
+				case nch = <-chs:
+				case <-ctx.Done():
+					return
+				}
```

for this and next selects, add a timeout.
it'll let the test fail faster if client is mis-behaving.